cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -1060,6 +1062,12 @@ private ApplicationSpecificationWithChangedWorkers(ApplicationSpecification dele
       return Transactions.execute(transactional, callable);
     } catch (TransactionFailureException e) {
       throw Transactions.propagate(e);
+    } catch (RuntimeException e) {
+      Throwable t = e.getCause();
+      if (t != null && t instanceof TException) {
```

Comment:
rather than put the logic in DefaultStore, I think it would be better to create a wrapper method similar to the existing Transactions.createTransactionalWithRetry method. This would make it reusable as we go through and modify system services that use transactions.

I suspect this behavior is always the correct behavior, in which case we would eventually move the logic from that wrapper method into CacheBasedTransactional.